main.vala

```diff
@@ -1,68 +1,64 @@
 using LanguageServer;
 using Gee;
 
-class Meson.TargetSource : Object {
-    public string language { get; set; }
-    public string[] compiler { get; set; }
-    public string[] parameters { get; set; }
-    public string[] sources { get; set; }
-}
-
-// currently unused
-class Meson.Target : Object {
-    public string name { get; set; }
-    public string id { get; set; }
-    public string defined_in { get; set; }
-    public string[] filename { get; set; }
-    public bool build_by_default { get; set; }
-}
-
-class Vls.TextDocument : Object {
-    private Context ctx;
-    private string filename;
-
-    public Vala.SourceFile file;
-    public string uri;
-    public int version;
-
-    public TextDocument (Context ctx,
-                         string filename,
-                         string? content = null,
-                         int version = 0) throws ConvertError, FileError {
-
-        if (!FileUtils.test (filename, FileTest.EXISTS)) {
-            throw new FileError.NOENT ("file %s does not exist".printf (filename));
-        }
+class Vls.Request {
+    private int64? int_value;
+    private string? string_value;
+    private string? method;
+
+    public Request (Variant id, string? method = null) {
+        assert (id.is_of_type (VariantType.INT64) || id.is_of_type (VariantType.STRING));
+        if (id.is_of_type (VariantType.INT64))
+            int_value = (int64) id;
+        else
+            string_value = (string) id;
+        this.method = method;
+    }
 
-        this.uri = Filename.to_uri (filename);
-        this.filename = filename;
-        this.version = version;
-        this.ctx = ctx;
+    public string to_string () {
+        string id_string = int_value != null ? int_value.to_string () : string_value;
+        return id_string + (method != null ? @":$method" : "");
+    }
 
-        var type = Vala.SourceFileType.NONE;
-        if (uri.has_suffix (".vala") || uri.has_suffix (".gs"))
-            type = Vala.SourceFileType.SOURCE;
-        else if (uri.has_suffix (".vapi"))
-            type = Vala.SourceFileType.PACKAGE;
+    public static uint hash (Request req) {
+        if (req.int_value != null)
+            return GLib.int64_hash (req.int_value);
+        else
+            return GLib.str_hash (req.string_value);
+    }
 
-        file = new Vala.SourceFile (ctx.code_context, type, filename, content);
-        if (type == Vala.SourceFileType.SOURCE) {
-            var ns_ref = new Vala.UsingDirective (new Vala.UnresolvedSymbol (null, "GLib", null));
-            file.add_using_directive (ns_ref);
-            ctx.add_using ("GLib");
+    public static bool equal (Request reqA, Request reqB) {
```

Comment:
To your first point, I don't think so. To your second point, because I need to pass this method to the HashSet containing Requests, which is the only place it's ever used, and this way I avoid having to pass a lambda to the HashSet constructor.